### PR TITLE
Allow PointerEnterEvent and PointerExitEvents to be created from any PointerEvent

### PR DESCRIPTION
**Pull request by ds84182**
## Description

When testing mouse hover support in a custom embedder, I noticed crashes in MouseTracker due to an implicit downcast from PointerEvent to PointerHoverEvent. The `_lastMouseEvent` map can contain PointerMoveEvents and PointerDownEvents, in addition to PointerHoverEvents, so changing the type of event handled by PointerEnter/ExitEvent.fromHoverEvent seems like the best course of action. Renaming the constructor would be a breaking change however.

Should I include tests for converting other pointer events into PointerEnter/ExitEvents?

## Related Issues

#29696

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]).

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by gspencergoog — COMMENTED** [review]
Thanks for fixing this! It's unfortunate that I didn't see this when I added multiple pointer support, I would have fixed the name too.  I don't think it's worth making a breaking change at this point, although I know that there are likely zero usages of that constructor in the wild.

You might add a test in `mouse_tracking_test.dart` that checks that mouse enter/exit works with mouse move events too. That will exercise this code, as well as the mouse tracking code.

**Comment by gspencergoog** [comment]
Actually, in discussing it more, we think it would be best to make the breaking change and call it `fromMouseEvent`.  There are literally no uses of this in open source software, and it's a very new API that is deep in internals, so it's unlikely to be used at all by third parties.

To do this, make a copy of your changed version of the constructor and mark the `fromHoverEvent` one as `@deprecated` and add a comment saying to use `fromMouseEvent` instead.

We'll need to make an announcement to the mailing list before committing, but I can do that.

**Comment by ds84182** [comment]
The test is going to be a bit more involved, unfortunately. `PointerHoverEventListener` only takes PointerHoverEvents. Should I create synthetic hover events like PointerEnterEvent does? Otherwise the test crashes [here](https://github.com/flutter/flutter/blob/e1e5b23ac6113c2967f8e06dc1d9f32889da243b/packages/flutter/lib/src/gestures/mouse_tracking.dart#L233).

**Comment by gspencergoog** [comment]
Hmm. I wonder if the line before that should be:
```dart
 if (hitAnnotation.annotation?.onHover != null && lastEvent is PointerHoverEvent) {
```
in any case. Move and Down events shouldn't be sent to `onHover` anyhow.

File: packages/flutter/lib/src/gestures/events.dart [hunk 1]
```diff
@@ -517,30 +517,35 @@ class PointerEnterEvent extends PointerEvent {
...
-    size: hover?.size,
-    radiusMajor: hover?.radiusMajor,
-    radiusMinor: hover?.radiusMinor,
-    radiusMin: hover?.radiusMin,
-    radiusMax: hover?.radiusMax,
-    orientation: hover?.orientation,
-    tilt: hover?.tilt,
-    synthesized: hover?.synthesized,
+  /// Deprecated. Please use [PointerEnterEvent.fromMouseEvent] instead.
+  @deprecated
```
Comment (goderbauer) [inline packages/flutter/lib/src/gestures/events.dart]: I believe `@Deprecated('use PointerEnterEvent.fromMouseEvent instead')` will generate slightly better analyzer warnings.

**Comment by ds84182** [comment]
Added a test

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by gspencergoog** [comment]
I'll land this as soon as the tree is green.

**Comment by gspencergoog** [comment]
Actually, because of the deprecations, we need to follow the "breaking change" protocols, so I'll send an announcement and we'll wait a bit for comments.

**Pull request merged**